lesson-3-sheduling-all-around/README.md

```diff
@@ -0,0 +1,266 @@
+kind: Pod
+metadata:
+  name: p1
+  labels:
+    security: S1
+spec:
+  containers:
+  - name: security-s1
+    image: docker.io/ocpqe/hello-pod
+```
+
+- p2 - which has affinity to be run on same node as "p1" using the selector be scheduled on same node as the pods with label "security" set to "S1"
+
+
+``` yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: p2
+  labels:
+    security: S1
+spec:
+  containers:
+  - name: security-s1
+    image: docker.io/ocpqe/hello-pod
+    podAffinity:
+      requiredDuringSchedulingIgnoredDuringExecution:
+      - labelSelector:
+          matchExpressions:
+          - key: security
+            operator: In
+            values:
+            - S1
+```
+
+- p2 - which has antiAffinity to NOT be run on same node as "p1" using the selector be scheduled on same node as the pods with label "security" set to "S1"
+
+``` yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: p3
+spec:
+  affinity:
+    podAntiAffinity:
+      requiredDuringSchedulingIgnoredDuringExecution:
+      - labelSelector:
+          matchExpressions:
+          - key: security
+            operator: In
+            values:
+            - S1
+        topologyKey: kubernetes.io/hostname
+  containers:
+  - name: pod-antiaffinity
+    image: docker.io/ocpqe/hello-pod
+```
+ 
+Now you can use `oc` command to double check if everything is running correctly.
+
+``` bash
+oc get pods -o jsonpath='{range .items[*]}{.metadata.name}{"\t"}{.spec.nodeName}{"\n"}'
+```
+
+As a task for yourself please create pod "p4" which will be scheduled on same node as pod "p3".
+You can also expand JSONPath filter to print the value of "security" label.
+
+If you are interested more in JSONPath you can find the docs at https://kubernetes.io/docs/reference/kubectl/jsonpath/
+
+
+To cleanup the environment we will run
+``` bash
+oc delete pod p1 p2 p3 p4
+```
+
+You can use these selector for more stuff. If you propagate your server placements, availability zones or HW equipment to labels. But be aware of bad resource utilization which can come with a complex rules and can even end up with upgrade issues.
+
+Also think about scalability. For example if you need two services to be running on the same node consider them being part of one pod - then they can scale easily together.
+
+
+Custome Scheduler
+=================
+
+Custom scheduler is a little bit overhead in a lot of cases. We are using it to prove that it is possible and to demonstrate how to access k8s api from pods. The complete source of the scheduler can be viewed in the `scheduler/app.py`. In this scenario we will:
+
+- setup OpenShift build for the scheduler app
+- build it
+- create service account and set th required permissions
+- run scheduler in OpenShift pod
+- Schedule a new pod using this scheduler
+
+
+So let's start with creating a binary build. This will enable us to build container in OpenShift using files on local filesystem. So run:
+``` bash
+oc new-build --binary --name=sched
```

Comment:
what a nonchalant way of introducing the attendees to a new openshift object :D

I can imagine there will be followup questions to builds